File: rust/2021/src/day23.rs
use aoc::dijkstra::DijkstraState;

// Stored as bytes rather than chars
// The whole state is then 27 bytes instead of 108,
// which makes hashing and copying far cheaper
type Rooms = [[u8; 4]; 4];
type Hallway = [u8; 11];

const fn hallway_index(room_index: usize) -> usize {
    room_index * 2 + 2
}

fn parse(filename: &str) -> Rooms {
    let lines = aoc::file::read_lines(filename);
    let top = lines[2].as_bytes();
    let bottom = lines[3].as_bytes();
    std::array::from_fn(|room_index| {
        // +1 for the leading # in the file line
        let col = hallway_index(room_index) + 1;
        [bottom[col], top[col], EMPTY, EMPTY]
    })
}

const HALLWAY: Hallway = *b"..X.X.X.X..";
const EMPTY: u8 = HALLWAY[0];
const FORBIDDEN: u8 = HALLWAY[2];

const fn numeric(c: u8) -> u8 {
    c - b'A'
}

const fn get_cost(c: u8) -> u32 {
    10_u32.pow(numeric(c) as u32)
}

const fn correct_room(c: u8, room_index: usize) -> bool {
    (numeric(c) as usize) == room_index
}

fn hallway_path(hallway: &Hallway, hall_index: usize, room_index: usize) -> Option<(usize, usize)> {
    let hall_above_room = hallway_index(room_index);
    let (from, to) = (
        hall_above_room.min(hall_index),
        hall_above_room.max(hall_index),
    );
    if (hallway[from..=to].iter().enumerate().any(|(i, c)| {
        return ((from + i) != hall_index) && (b'A'..=b'D').contains(c);
    })) {
        // There is another amphipod in the way
        return None;
    }
    return Some((from, to));
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
struct Configuration {
    hallway: Hallway,
    rooms: Rooms,
}
impl std::hash::Hash for Configuration {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        // The derived impl feeds the rooms to the hasher one row at a time
        state.write(&self.hallway);
        state.write(self.rooms.as_flattened());
    }
}

/// Admissible lower bound on the cost of finishing from `config`:
/// every amphipod that is not already settled has to pay at least
/// its own way out of the room, along the hallway and back down.
fn heuristic(config: &Configuration, room_size: usize) -> u32 {
    let mut cost = 0_u32;

    for (h, &c) in config.hallway.iter().enumerate() {
        if !(b'A'..=b'D').contains(&c) {
            continue;
        }
        let target_room = numeric(c) as usize;
        let steps = h.abs_diff(hallway_index(target_room)) as u32 + 1;
        cost += steps * get_cost(c);
    }

    for (r, room) in config.rooms.iter().enumerate() {
        for s in 0..room_size {
            let c = room[s];
            if c == EMPTY {
                continue;
            }
            let target_room = numeric(c) as usize;

            if (target_room == r) {
                // Settled if all slots below are also in the correct room
                if (0..s).all(|b| room[b] != EMPTY && correct_room(room[b], r)) {
                    continue;
                }
                // Must exit and re-enter (wrong amphipod is stuck below)
                cost += ((room_size - s) as u32 + 1) * get_cost(c);
            } else {
                let steps_out = (room_size - s) as u32;
                let h_steps = hallway_index(r).abs_diff(hallway_index(target_room)) as u32;
                cost += (steps_out + h_steps + 1) * get_cost(c);
            }
        }
    }

    return cost;
}

fn solve(rooms: &Rooms, room_size: usize) -> u32 {
    let start = Configuration {
        hallway: HALLWAY,
        rooms: *rooms,
    };
    let mut end_rooms = [[EMPTY; 4]; 4];
    for room_index in 0..4 {
        for slot in 0..room_size {
            end_rooms[room_index][slot] = b'A' + room_index as u8;
        }
    }
    let end = Configuration {
        hallway: HALLWAY,
        rooms: end_rooms,
    };
    let distances = aoc::dijkstra::shortest_distances_astar(
        &start,
        |current| *current == end,
        |current| {
            let mut neighbors = Vec::new();

            // Try moving from room to the hallway
            for (room_index, room) in current.rooms.iter().enumerate() {
                let Some(top_slot) = (0..room_size).rev().find(|&i| room[i] != EMPTY)
                //
                else {
                    continue;
                };
                if (0..=top_slot).all(|i| correct_room(room[i], room_index)) {
                    continue; // Room settled
                }
                let amphipod = room[top_slot];
                let steps_up = (room_size - top_slot) as u32;

                let mut new_room = *room;
                new_room[top_slot] = EMPTY;

                for (hall_index, hall) in current.hallway.iter().enumerate() {
                    if (*hall != EMPTY) {
                        continue;
                    }
                    let Some(path) = hallway_path(&current.hallway, hall_index, room_index)
                    //
                    else {
                        continue;
                    };

                    let distance = (steps_up + (path.1 - path.0) as u32);

                    let mut new_hallway = current.hallway;
                    new_hallway[hall_index] = amphipod;

                    let mut new_rooms = current.rooms;
                    new_rooms[room_index] = new_room;

                    neighbors.push(DijkstraState {
                        data: Configuration {
                            hallway: new_hallway,
                            rooms: new_rooms,
                        },
                        distance: distance * get_cost(amphipod),
                    });
                }
            }

            // Try moving from the hallway to a room
            for (hall_index, hall) in current.hallway.iter().enumerate() {
                if ((*hall == EMPTY) || (*hall == FORBIDDEN)) {
                    continue;
                }

                let amphipod = *hall;
                let room_index = numeric(amphipod) as usize;
                let room = &current.rooms[room_index];

                if (0..room_size).any(|i| room[i] != EMPTY && !correct_room(room[i], room_index)) {
                    continue; // Wrong amphipod in room
                }
                let Some(target) = (0..room_size).find(|&i| room[i] == EMPTY)
                //
                else {
                    continue;
                };
                let Some(path) = hallway_path(&current.hallway, hall_index, room_index)
                //
                else {
                    continue;
                };

                let steps_down = (room_size - target) as u32;
                let distance = (steps_down + (path.1 - path.0) as u32);

                let mut new_hallway = current.hallway;
                new_hallway[hall_index] = EMPTY;

                let mut new_rooms = current.rooms;
                new_rooms[room_index][target] = amphipod;

                neighbors.push(DijkstraState {
                    data: Configuration {
                        hallway: new_hallway,
                        rooms: new_rooms,
                    },
                    distance: distance * get_cost(amphipod),
                });
            }

            return neighbors;
        },
        |current| heuristic(current, room_size),
    );

    return distances[&end];
}

fn solve_case1(rooms: &Rooms) -> u32 {
    solve(rooms, 2)
}

fn solve_case2(rooms: &Rooms) -> u32 {
    // #D#C#B#A# -> slot 2
    // #D#B#A#C# -> slot 1
    let inserted_upper = *b"DCBA";
    let inserted_lower = *b"DBAC";
    let rooms4 = std::array::from_fn(|i| {
        [
            rooms[i][0],
            inserted_lower[i],
            inserted_upper[i],
            rooms[i][1],
        ]
    });
    solve(&rooms4, 4)
}

fn main() {
    println!("Part 1");
    let example = parse("day23.example");
    aoc::expect_result!(12521, solve_case1(&example));
    let input = parse("day23.input");
    aoc::expect_result!(14350, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(44169, solve_case2(&example));
    aoc::expect_result!(49742, solve_case2(&input));
}

File: 2021/day23.cpp
// https://adventofcode.com/2021/day/23

#include "../common/common.h"
#include "../common/rust.h"

#include <array>
#include <optional>
#include <print>
#include <string>
#include <vector>

using Hallway = std::array<char, 11>;
using Rooms = std::array<std::array<char, 4>, 4>;
using Input = Rooms;

constexpr let HALLWAY =
    Hallway{'.', '.', 'X', '.', 'X', '.', 'X', '.', 'X', '.', '.'};
constexpr let EMPTY = HALLWAY[0];
constexpr let FORBIDDEN = HALLWAY[2];

fn hallway_index(usize room_index) -> usize { return room_index * 2 + 2; }

fn parse(String const& filename) -> Input {
  let lines = aoc::views::read_lines(filename, aoc::keep_spaces{}) |
              aoc::ranges::to<Vec<String>>();
  Input rooms;
  for (let room_index : Range{0uz, 4uz}) {
    let col = hallway_index(room_index) + 1;
    rooms[room_index] = {lines[3][col], lines[2][col], EMPTY, EMPTY};
  }
  return rooms;
}

struct Configuration {
  Hallway hallway;
  Rooms rooms;
  fn operator<=>(const Configuration&) const = default;
};

fn get_cost(char c) -> u32 {
  constexpr std::array<u32, 4> costs = {1, 10, 100, 1000};
  return costs[static_cast<usize>(c - 'A')];
}

fn correct_room(char c, usize room_index) -> bool {
  return static_cast<usize>(c - 'A') == room_index;
}

fn hallway_path(Hallway const& hallway, usize hall_index, usize room_index)
    -> Option<std::pair<usize, usize>> {
  let hall_above_room = hallway_index(room_index);
  let from = std::min(hall_above_room, hall_index);
  let to = std::max(hall_above_room, hall_index);
  for (let i : Range{from, to + 1}) {
    if ((i != hall_index) && (hallway[i] >= 'A') && (hallway[i] <= 'D')) {
      return None;
    }
  }
  return std::pair{from, to};
}

fn heuristic(Configuration const& config, usize room_size) -> i32 {
  auto cost = i32{};

  for (let h : Range{0uz, config.hallway.size()}) {
    let c = config.hallway[h];
    if ((c < 'A') || (c > 'D')) {
      continue;
    }
    let target_room = static_cast<usize>(c - 'A');
    let steps = std::abs(static_cast<i32>(h) -
                         static_cast<i32>(hallway_index(target_room))) +
                1;
    cost += steps * static_cast<i32>(get_cost(c));
  }

  for (let r : Range{0uz, 4uz}) {
    for (let s : Range{0uz, room_size}) {
      let c = config.rooms[r][s];
      if (c == EMPTY) {
        continue;
      }
      let target_room = static_cast<usize>(c - 'A');

      if (target_room == r) {
        // Settled if all slots below are also in the correct room
        bool below_ok = true;
        for (usize b = 0; b < s; ++b) {
          if (config.rooms[r][b] == EMPTY ||
              static_cast<usize>(config.rooms[r][b] - 'A') != r) {
            below_ok = false;
            break;
          }
        }
        if (below_ok)
          continue;
        // Must exit and re-enter (wrong amphipod is stuck below)
        let steps = static_cast<i32>(room_size - s) + 1;
        cost += steps * static_cast<i32>(get_cost(c));
      } else {
        let steps_out = static_cast<i32>(room_size - s);
        let h_steps = std::abs(static_cast<i32>(hallway_index(r)) -
                               static_cast<i32>(hallway_index(target_room)));
        cost += (steps_out + h_steps + 1) * static_cast<i32>(get_cost(c));
      }
    }
  }

  return cost;
}

fn solve(Rooms const& rooms, usize room_size) -> u32 {
  let start = Configuration{HALLWAY, rooms};

  auto end_rooms = Rooms{};
  for (auto& room : end_rooms) {
    room.fill(EMPTY);
  }
  for (usize room_index = 0; room_index < 4; ++room_index) {
    for (usize slot = 0; slot < room_size; ++slot) {
      end_rooms[room_index][slot] = static_cast<char>('A' + room_index);
    }
  }
  let end = Configuration{HALLWAY, std::move(end_rooms)};

  let distances = aoc::shortest_distances_astar(
      start,
      [&](Configuration const& current) {
        auto neighbors = Vec<aoc::dijkstra_neighbor_t<Configuration>>{};

        // Try moving from room to the hallway
        for (usize room_index = 0; room_index < 4; ++room_index) {
          let& room = current.rooms[room_index];

          // Find topmost occupied slot within room_size
          i32 top_slot = -1;
          for (i32 i = static_cast<i32>(room_size) - 1; i >= 0; --i) {
            if (room[i] != EMPTY) {
              top_slot = i;
              break;
            }
          }
          if (top_slot < 0) {
            continue;
          }

          // Skip if all occupied slots are already correct
          bool settled = true;
          for (u32 i = 0; i <= top_slot; ++i) {
            if (!correct_room(room[i], room_index)) {
              settled = false;
              break;
            }
          }
          if (settled) {
            continue;
          }

          let amphipod = room[top_slot];
          let steps_up = static_cast<i32>(room_size) - top_slot;
          auto new_room = room;
          new_room[top_slot] = EMPTY;

          for (usize hall_index = 0; hall_index < current.hallway.size();
               ++hall_index) {
            if (current.hallway[hall_index] != EMPTY) {
              continue;
            }
            let path = hallway_path(current.hallway, hall_index, room_index);
            if (!path) {
              continue;
            }
            let distance =
                (steps_up + static_cast<u32>(path->second - path->first)) *
                get_cost(amphipod);

            auto new_hallway = current.hallway;
            new_hallway[hall_index] = amphipod;

            auto new_rooms = current.rooms;
            new_rooms[room_index] = new_room;

            neighbors.emplace_back(Configuration{new_hallway, new_rooms},
                                   distance);
          }
        }

        // Try moving from the hallway to a room
        for (usize hall_index = 0; hall_index < current.hallway.size();
             ++hall_index) {
          let hall = current.hallway[hall_index];
          if ((hall == EMPTY) || (hall == FORBIDDEN)) {
            continue;
          }
          let amphipod = hall;
          let room_index = static_cast<usize>(amphipod - 'A');
          let& room = current.rooms[room_index];

          // Skip if any wrong amphipod is in the room
          bool has_wrong = false;
          for (usize i = 0; i < room_size; ++i) {
            if ((room[i] != EMPTY) && !correct_room(room[i], room_index)) {
              has_wrong = true;
              break;
            }
          }
          if (has_wrong) {
            continue;
          }

          // Find the deepest empty slot
          i32 target = -1;
          for (usize i = 0; i < room_size; ++i) {
            if (room[i] == EMPTY) {
              target = static_cast<i32>(i);
              break;
            }
          }
          if (target < 0) {
            continue;
          }

          let path = hallway_path(current.hallway, hall_index, room_index);
          if (!path) {
            continue;
          }
          let steps_down = static_cast<i32>(room_size) - target;
          let distance =
              (steps_down + static_cast<u32>(path->second - path->first)) *
              get_cost(amphipod);

          auto new_hallway = current.hallway;
          new_hallway[hall_index] = EMPTY;

          auto new_rooms = current.rooms;
          new_rooms[room_index][target] = amphipod;

          neighbors.emplace_back(Configuration{new_hallway, new_rooms},
                                 distance);
        }

        return neighbors;
      },
      [&](Configuration const& config) { return heuristic(config, room_size); },
      end);

  return distances.find(end)->second;
}

fn solve_case1(Rooms const& rooms) -> u32 { return solve(rooms, 2); }

fn solve_case2(Rooms const& rooms) -> u32 {
  //   #D#C#B#A#  -> slot 2
  //   #D#B#A#C#  -> slot 1
  constexpr std::array<char, 4> inserted_upper = {'D', 'C', 'B', 'A'};
  constexpr std::array<char, 4> inserted_lower = {'D', 'B', 'A', 'C'};
  Rooms rooms4;
  for (usize i = 0; i < 4; ++i) {
    rooms4[i] = {rooms[i][0], inserted_lower[i], inserted_upper[i],
                 rooms[i][1]};
  }
  return solve(rooms4, 4);
}

int main() {
  std::println("Part 1");
  let example = parse("day23.example");
  AOC_EXPECT_RESULT(12521, solve_case1(example));
  let input = parse("day23.input");
  AOC_EXPECT_RESULT(14350, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(44169, solve_case2(example));
  AOC_EXPECT_RESULT(49742, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}
